honeybadgermpc/field.py

```diff
@@ -94,246 +96,268 @@ def __int__(self):
-            return (lastx, lasty, a)
-
-        inverse = extended_gcd(self.value, self.modulus)[0]
-        return GFElement(inverse, self.field)
-
-    def __div__(self, other):
-        """Division."""
-        try:
-            if self.field is not other.field:
-                raise FieldsNotIdentical
-            return self * ~other
-        except AttributeError:
-            return self * ~GFElement(other, self.field)
-
-    __truediv__ = __div__
-    __floordiv__ = __div__
-
-    def __rdiv__(self, other):
-        """Division (reflected argument version)."""
-        return GFElement(other, self.field) / self
-
-    __rtruediv__ = __rdiv__
-    __rfloordiv__ = __rdiv__
-
-    def sqrt(self):
-        """Square root.
-
-        No attempt is made the to return the positive square root.
-
-        Computing square roots is only possible when the modulus
-        is a Blum prime (congruent to 3 mod 4).
-        """
-        assert self.modulus % 4 == 3, "Cannot compute square " \
-            "root of %s with modulus %s" % (self, self.modulus)
+def makeGF(modulus, val):
+    return GF(modulus)(val)
 
-        # Because we assert that the modulus is a Blum prime
-        # (congruent to 3 mod 4), there will be no reminder in the
-        # division below.
-        root = pow(self.value, (self.modulus+1)//4, self.modulus)
-        return GFElement(root, self.field)
 
-    def bit(self, index):
-        """Extract a bit (index is counted from zero)."""
-        return (self.value >> index) & 1
+def GF(modulus):
+    """Generate a Galois (finite) field with the given modulus.
 
-    def signed(self):
-        """Return a signed integer representation of the value.
+    The modulus must be a prime:
 
-        If x > floor(p/2) then subtract p to obtain negative integer.
-        """
-        if self.value > ((self.modulus-1)/2):
-            return self.value - self.modulus
-        else:
+    >>> Z23 = GF(23) # works
+    >>> Z10 = GF(10) # not a prime
+    Traceback (most recent call last):
+        ...
+    ValueError: 10 is not a prime
+
+    A modulus of 256 is special since it returns the GF(2^8) field
+    even though 256 is no prime:
+
+    >>> GF256 = GF(256)
+    >>> print GF256(1)
+    [1]
+
+    Please note, that if you wish to calculate square roots, the
+    modulus must be a Blum prime (congruent to 3 mod 4):
+
+    >>> Z17 = GF(17) # 17 % 4 == 1, so 17 is no Blum prime
+    >>> x = Z17(10)
+    >>> x.sqrt()
+    Traceback (most recent call last):
+        ...
+    AssertionError: Cannot compute square root of {10} with modulus 17
+    """
+    if modulus in _field_cache:
+        return _field_cache[modulus]
+
+    if not mpz(modulus).is_prime():
+        raise ValueError("%d is not a prime" % modulus)
+
+    # Define a new class representing the field. This class will be
+    # returned at the end of the function.
+    class GFElement(FieldElement):
+
+        def __reduce__(self):
```

Comment:
I had tried something similar, I was still getting the "cannot pickle local object" error so I assumed that a local class can't be reduced using `pickle`. oO